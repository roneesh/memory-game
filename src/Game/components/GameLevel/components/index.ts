import LevelReady from './LevelReady';
// import LevelPlaying from './LevelPlaying';
// import LevelFinished from './LevelFinished';
// import LevelPreview from './LevelPreview';

export {
  LevelReady,
  // LevelPlaying,
  // LevelFinished,
  // LevelPreview,
};
