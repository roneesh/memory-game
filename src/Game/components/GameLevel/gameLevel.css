.game-level {
  margin-bottom: 20px;
}

.game-board {
  background-color: #4f5d75;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
  padding: 3px 0;
}

.game-board__tile {
  width: 100px;
  height: 100px;
  margin: 3px;
  perspective: 300px;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: #bfc0c0;
  color: #ffffff;
  font-size: 32px;
  font-family: 'Monoton', cursive;
}

@keyframes tileWiggle {
  0%   { transform:  rotateZ(0deg) }
  25%  { transform:  rotateZ(5deg) }
  50%  { transform:  rotateZ(-5deg) }
  100% { transform:  rotateZ(0deg) }
}

.game-board__tile--wiggling {
  position: relative;
  animation-name: tileWiggle;
  animation-duration: 3s;
}

@keyframes tileExplodeUp {
  0%   { top: 0px }
  15%  { top: 100px }
  40%  { top: 1000px }
  100% { top: 10000px }
}

@keyframes tileExplodeDown {
  0%   { top: 0px }
  15%  { top: -100px }
  40%  { top: -1000px }
  100% { top: -10000px }
}

.game-board__tile--exploding-up {
  position: relative;
  animation-name: tileExplodeUp;
  animation-duration: 3s;
}

.game-board__tile--exploding-down {
  position: relative;
  animation-name: tileExplodeDown;
  animation-duration: 3s;
}
