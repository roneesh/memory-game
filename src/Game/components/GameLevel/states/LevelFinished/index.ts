import LevelFinished from './LevelFinished';

export default LevelFinished;
