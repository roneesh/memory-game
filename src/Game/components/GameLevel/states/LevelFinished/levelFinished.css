.game-board__tile--expldoing {

}