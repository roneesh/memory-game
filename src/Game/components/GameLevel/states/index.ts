// import LevelFinished from './LevelFinished';
// import LevelPlaying from './LevelPlaying';
import LevelPreview from './LevelPreview';
import LevelReady from './LevelReady';

export {
  // LevelFinished,
  // LevelPlaying,
  LevelPreview,
  LevelReady,
};
