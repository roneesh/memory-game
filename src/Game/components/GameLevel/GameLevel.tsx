import { shuffle } from 'lodash';
import * as React from 'react';
import './gameLevel.css';
import { LevelFinished, LevelPlaying, LevelPreview, LevelReady, } from './states';

interface IGameLevelProps {
  updateLevel: () => void;
  level: number;
};

type LevelState = "ready" | "preview" | "playing" | "finished";
export interface IBoardTile {
  character: string | HTMLElement;
};
export interface ILevelTurns {
  count: number;
  currentTurn: number[];
}
interface IGameLevelState {
  levelState: LevelState;
  board: IBoardTile[];
  turns:  ILevelTurns;
};

class GameLevel extends React.Component<IGameLevelProps, IGameLevelState> {
  constructor(props: IGameLevelProps) {
    super(props);

    this.state = {
      board: this.generateBoard(),
      levelState: 'ready',
      turns: {
        count: 0,
        currentTurn: [],
      },
    };
  }

  public generateBoard(): IBoardTile[] {
    return shuffle(
      ['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D', 'E', 'E', 'F', 'F',
      'G', 'G', 'H', 'H', 'I', 'I', 'J', 'J', 'K', 'K', 'L', 'L'].map(character => {
        return {
          character,
        };
      })
    );
  }

  public setLevelState = (newLevelState: LevelState) => {
    if (newLevelState === 'ready') {
      this.setState({
        board: this.generateBoard(),
        levelState: newLevelState,
        turns: {
          count: 0,
          currentTurn: [],
        },
      });
    }
    this.setState({ levelState: newLevelState });
  }

  public getResetBtn() {
    const { levelState } = this.state;

    if (levelState === 'finished') {
      return null;
    }

    const text = levelState === 'ready' ? 'Play Level' : 'Reset';
    const nextState = levelState === 'ready' ? 'preview' : 'ready';
    return <button onClick={this.setLevelState.bind(null, nextState)}>
      {text}
    </button>;
  }

  public renderControls() {
    return <div className="game-level__controls">
      {this.getResetBtn()}
    </div>
  }

  public renderState() {
    const { levelState, board } = this.state;
    const { updateLevel } = this.props;

    switch(levelState) {
      case 'ready':
        return <LevelReady board={board} />;
      case 'preview':
        return <LevelPreview board={board} setLevelState={this.setLevelState} />;
      case 'playing':
        return <LevelPlaying board={board} setLevelState={this.setLevelState} />;
      case 'finished':
        return <LevelFinished board={board} updateLevel={updateLevel} />;
      default:
        return null;
    }
  }

  public componentDidUpdate(prevProps: IGameLevelProps) {
    if (this.props.level !== prevProps.level) {
      this.setLevelState('ready');
    }
  }
  public render() {
    return <div className="game-level">
      {this.renderControls()}
      {this.renderState()}
    </div>;
  }
}

export default GameLevel;
