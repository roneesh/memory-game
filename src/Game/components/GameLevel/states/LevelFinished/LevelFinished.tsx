import * as React from 'react';
import { IBoardTile } from '../../GameLevel';

interface ILevelFinishedProps {
  board: IBoardTile[];
  updateLevel: () => void;
};

interface ILevelFinishedState {
  finishedTimer: NodeJS.Timer;
  timeElapsed: number;
  action: 'wiggling' | 'exploding';
};

class LevelFinished extends React.Component<ILevelFinishedProps, ILevelFinishedState> {
  public timeIncrement = 1000;

  constructor(props: ILevelFinishedProps) {
    super(props);

    this.state = {
      action: 'wiggling',
      finishedTimer: setInterval(this.finishedManager, this.timeIncrement),
      timeElapsed: 0,
    };
  }

  public componentWillUnmount() {
    clearInterval(this.state.finishedTimer);
  }

  public finishedManager = () => {
    const { updateLevel } = this.props;
    const { timeElapsed } = this.state;
    if (timeElapsed < 3000) {
      this.setState({
        timeElapsed: timeElapsed + this.timeIncrement,
      });
      return;
    }
    if (timeElapsed < 6000) {
      this.setState({
        action: 'exploding',
        timeElapsed: timeElapsed + this.timeIncrement,
      });
      return;
    }
    updateLevel();
  }

  public render() {
    const { board } = this.props;
    const { action } = this.state;

    return <div className="game-board">
      {board.map((tile, index) => {
        const movementClass =
          action === 'wiggling' ? 'wiggling' : index <= 11 ? 'exploding-up' : 'exploding-down';
        const tileClass = `game-board__tile game-board__tile--${movementClass}`;

        return <div className={tileClass} key={index}>
          {tile.character}
      </div>;
      })}
    </div>
  }

};

export default LevelFinished;
