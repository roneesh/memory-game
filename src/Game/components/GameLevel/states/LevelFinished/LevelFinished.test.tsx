import { shallow } from 'enzyme';
import * as React from 'react';
import renderer from 'react-test-renderer';
import { IBoardTile } from '../../GameLevel';
import LevelFinished from './LevelFinished';

describe('The level when its previewing', () => {
  let board: IBoardTile[] = [];
  let updateLevelSpy: any;
  const boardBase =
    ['A', 'A', 'B', 'B', 'C', 'C', 'D', 'D', 'E', 'E', 'F', 'F',
    'G', 'G', 'H', 'H', 'I', 'I', 'J', 'J', 'K', 'K', 'L', 'L'];

  const getComponent = () => shallow((<LevelFinished board={board} updateLevel={updateLevelSpy} />));
  const getSnapshot = () => <LevelFinished board={board} updateLevel={jest.fn()} />;

  beforeEach(() => {
    updateLevelSpy = jest.fn();
    board = boardBase.map(character => {
      return {
        character,
      };
    });
  });

  it('it renders as wiggling tiles', () => {
    const levelFinished = renderer
      .create(getSnapshot())
      .toJSON();
    expect(levelFinished).toMatchSnapshot();
  });

  it('turns wiggling tiles into exploding ones', () => {
    const component = getComponent();
    jest.runTimersToTime(6000);

    expect(component.find('.game-board__tile--exploding-up').length).toEqual(12);
    expect(component.find('.game-board__tile--exploding-down').length).toEqual(12);
  });

  it('dispatches a game update after 9000ms', () => {
    getComponent();
    jest.runTimersToTime(9000);
    expect(updateLevelSpy).toHaveBeenCalled();
  });

});
