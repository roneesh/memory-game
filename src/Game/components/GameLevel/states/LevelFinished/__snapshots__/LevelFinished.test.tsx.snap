// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The level when its previewing it renders as wiggling tiles 1`] = `
<div
  className="game-board"
>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    A
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    A
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    B
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    B
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    C
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    C
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    D
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    D
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    E
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    E
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    F
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    F
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    G
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    G
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    H
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    H
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    I
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    I
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    J
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    J
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    K
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    K
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    L
  </div>
  <div
    className="game-board__tile game-board__tile--wiggling"
  >
    L
  </div>
</div>
`;
